README.md
<p align="center">
    <h1 align="center">The Blibox</h1>
</p>

![Blibox](docs/img/banner.png)

**[Usage](#usage)** |
**[Architecture](#architecture)** |
**[Community](#community)** |
**[Features](#feature-overview)** |
**[Intranet](#intranet-overview)** |
**[Screenshots](#screenshots)** |
**[Contributing](#contributing-)** |
**[Logos](#logos)** |
**[License](#license)**

![Blibox](docs/_includes/figures/blibox/blibox-intranet-dash-all.png)

[![Release](https://img.shields.io/github/release/cytopia/blibox.svg?colorB=orange)](https://github.com/cytopia/blibox/releases)
[![Gitter](https://badges.gitter.im/blibox/Lobby.svg)](https://gitter.im/blibox/Lobby?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
[![Discourse](https://img.shields.io/discourse/https/blibox.discourse.group/status.svg?colorB=%234CB697)](https://blibox.discourse.group)
[![type](https://img.shields.io/badge/type-Docker-blue.svg)](https://www.docker.com/)
[![License](https://img.shields.io/badge/license-MIT-%233DA639.svg)](https://opensource.org/licenses/MIT)

[![Build Status](https://travis-ci.org/cytopia/blibox.svg?branch=master)](https://travis-ci.org/cytopia/blibox)
[![Documentation Status](https://readthedocs.org/projects/blibox/badge/?version=latest)](https://blibox.readthedocs.io)
[![Build Status](https://github.com/cytopia/blibox/workflows/Lint/badge.svg)](https://github.com/cytopia/blibox/actions?workflow=Lint)
[![Build Status](https://github.com/cytopia/blibox/workflows/Docs/badge.svg)](https://github.com/cytopia/blibox/actions?workflow=Docs)
[![Build Status](https://github.com/cytopia/blibox/workflows/Linux/badge.svg)](https://github.com/cytopia/blibox/actions?workflow=Linux)
[![Build Status](https://github.com/cytopia/blibox/workflows/Config/badge.svg)](https://github.com/cytopia/blibox/actions?workflow=Config)
[![Build Status](https://github.com/cytopia/blibox/workflows/Versions/badge.svg)](https://github.com/cytopia/blibox/actions?workflow=Versions)

<img width="200" style="width:200px;" src="docs/_includes/figures/https/https-ssl-address-bar.png" /><br/>
<small><sub>Support for <a href="https://blibox.readthedocs.io/en/latest/intermediate/setup-valid-https.html">valid https</a> out of the box.</sub></small>

The Blibox is a modern and highly customisable **dockerized PHP stack** supporting full **LAMP**
and **MEAN** and running on all major platforms.  The main goal is to easily switch and combine
any version required for local development. It supports an **unlimited number of projects** for
which **vhosts**, **SSL certificates** and **DNS records** are created automatically.
**Reverse proxies** per project are supported to ensure listening server such as NodeJS can also be reached.
Email catch-all and popular development tools will be at your service as well. Configuration is not necessary, as everything is already pre-setup.

Furthermore, the Blibox provides an **identical** and **reproducible development environment** for different host operating systems.

**Requirements**

![Linux](https://raw.githubusercontent.com/cytopia/icons/master/64x64/linux.png)
![OSX](https://raw.githubusercontent.com/cytopia/icons/master/64x64/osx.png)
![Windows](https://raw.githubusercontent.com/cytopia/icons/master/64x64/windows.png)
![Plus](https://raw.githubusercontent.com/cytopia/icons/master/64x64/plus.png)
![Docker](https://raw.githubusercontent.com/cytopia/icons/master/64x64/docker.png)

* [Docker Engine 17.06.0+](https://docs.docker.com/compose/compose-file/compose-versioning/#version-23)
* [Docker Compose 1.16.0+](https://docs.docker.com/compose/compose-file/compose-versioning/#version-23)

## Architecture

#### Available Stacks

The Blibox aims to be a swiss army knife for local development by providing you all the services
you would ever need. To get an idea about the architecture behind it and to also see what's available
have a look at the following diagrams and tables.

<table width="100%" style="width:100%;display:table">
 <tr>
  <td width="30%" valign="top" style="width:30%; vertical-align:top;">
   <h4>Smallest stack</h4>
   <p>This is the smallest possible and fully functional stack you can run</p>
  </td>
  <td width="70%" valign="top" style="width:70%; vertical-align:top;">
   <h4>Full stack</h4>
   <p>To better understand what is actually possible have a look at the full example</p>
  </td>
 </tr>
 <tr>
  <td>
   <img width="300" style="width:300px" title="Blibox stack" src="https://raw.githubusercontent.com/blibox/artwork/master/submissions_diagrams/cytopia/02/png/architecture-small.png" />
  </td>
  <td>
   <img title="Blibox stack" src="https://raw.githubusercontent.com/blibox/artwork/master/submissions_diagrams/cytopia/01/png/architecture-full.png" />
  </td>
 </tr>
</table>

> [Blibox artwork](https://github.com/blibox/artwork)

#### Available Container

The following table lists all integrated and pre-configured Docker container shipped by the Blibox.
Only the webserver and PHP container are mandatory, all others are optional and don't need to be started.

Each of them is also available in multiple different versions in order to reflect your exact desired environment.

| Accel   | Web        | App            | SQL        | NoSQL     | Queue / Search | ELK           | Utils     |
|---------|------------|----------------|------------|-----------|----------------|---------------|-----------|
| HAProxy | Apache     | PHP            | MariaDB    | Memcached | RabbitMQ       | ElasticSearch | Bind      |
| Varnish | Nginx      | Python (Flask) | MySQL      | MongoDB   | Solr           | Logstash      | Blackfire |
|         |            |                | PerconaDB  | Redis     |                | Kibana        | MailHog   |
|         |            |                | PostgreSQL |           |                |               | Ngrok     |

> **Documentation:**
> [Available Container](https://blibox.readthedocs.io/en/latest/readings/available-container.html)

## Community

The Blibox has a lot of features reaching from a simple single-user development environment that
works out of the box up to a shared development infrastructure for a corporate network.

In order to be aware about all that features, ensure to have skimmed over the
**[documentation](https://blibox.readthedocs.io)**, so you know what can be done and how that might
simplify your every-day life. If you ever run into any unforseen issues, feel free to join the
**[chat](https://gitter.im/blibox/Lobby)** or visit the **[forums](https://blibox.discourse.group)** and get community support quickly.

<table width="100%" style="width:100%; display:table;">
 <thead>
  <tr>
   <th width="25%" style="width:25%;"><h3><a target="_blank" href="https://blibox.readthedocs.io">Documentation</a></h3></th>
   <th width="25%" style="width:25%;"><h3><a target="_blank" href="https://gitter.im/blibox/Lobby">Chat</a></h3></th>
   <th width="25%" style="width:25%;"><h3><a target="_blank" href="https://blibox.discourse.group">Forum</a></h3></th>
   <th width="25%" style="width:25%;"><h3><a target="_blank" href="https://github.com/blibox/flames">Flames</a></h3></th>
  </tr>
 </thead>
 <tbody style="vertical-align: middle; text-align: center;">
  <tr>
   <td>
    <a target="_blank" href="https://blibox.readthedocs.io">
     <img title="Documentation" name="Documentation" src="https://raw.githubusercontent.com/cytopia/icons/master/400x400/readthedocs.png" />
    </a>
   </td>
   <td>
    <a target="_blank" href="https://gitter.im/blibox/Lobby">
     <img title="Chat on Gitter" name="Chat on Gitter" src="https://raw.githubusercontent.com/cytopia/icons/master/400x400/gitter.png" />
    </a>
   </td>
   <td>
    <a target="_blank" href="https://blibox.discourse.group">
     <img title="Blibox Forums" name="Forum" src="https://raw.githubusercontent.com/cytopia/icons/master/400x400/discourse.png" />
    </a>
   </td>
   <td>
    <a target="_blank" href="https://github.com/blibox/flames">
     <img title="Blibox Flames" name="Flames" src="https://raw.githubusercontent.com/cytopia/icons/master/400x400/flames2.png" />
    </a>
   </td>
  </tr>
  <tr>
  <td><a target="_blank" href="https://blibox.readthedocs.io">blibox.readthedocs.io</a></td>
  <td><a target="_blank" href="https://gitter.im/blibox/Lobby">gitter.im/blibox</a></td>
  <td><a target="_blank" href="https://blibox.discourse.group">blibox.discourse.group</a></td>
  <td><a target="_blank" href="https://github.com/blibox/flames">github.com/blibox/flames</a></td>
  </tr>
 </tbody>
</table>

## Usage

#### Quick start

<table width="100%" style="width:100%; display:table;">
 <thead>
  <tr>
   <th width="50%" style="width:50%;">Linux and MacOS</th>
   <th width="50%" style="width:50%;">Windows</th>
  </tr>
 </thead>
 <tbody style="vertical-align: bottom;">
  <tr>
   <td>
<div class="highlight highlight-source-shell"><pre># Get the Blibox
git clone https://github.com/cytopia/blibox</pre></div>
<div class="highlight highlight-source-shell"><pre># Create docker-compose environment file
cd blibox
cp env-example .env</pre></div>
<div class="highlight highlight-source-shell"><pre># Edit your configuration
vim .env</pre></div>
<div class="highlight highlight-source-shell"><pre># Start all container
docker-compose up</pre></div>
   </td>
   <td>
    1. Clone <code>https://github.com/cytopia/blibox</code> to <code>C:\blibox</code> with <a href="https://git-scm.com/downloads">Git for Windows</a><br/><br/>
    2. Copy <code>C:\blibox\env-example</code> to <code>C:\blibox\.env</code><br/><br/>
    3. Edit <code>C:\blibox\.env</code><br/><br/>
    4. <a href="https://blibox.readthedocs.io/en/latest/howto/terminal/open-terminal-on-win.html">Open a terminal on Windows</a> and type:<br/><br/><br/>
    <pre># Start all container
C:\blibox> docker-compose up</pre></div>
   </td>
  </tr>
 </tbody>
</table>

> **Documentation:**
> [Install the Blibox](https://blibox.readthedocs.io/en/latest/getting-started/install-the-blibox.html) |
> [Start the Blibox](https://blibox.readthedocs.io/en/latest/getting-started/start-the-blibox.html) |
> [.env file](https://blibox.readthedocs.io/en/latest/configuration-files/env-file.html)

#### Selective start

The above will start all containers, you can however also just start the containers you actually need. This is achieved by simply specifying them in the docker-compose command.

```bash
docker-compose up httpd php mysql redis
```
> **Documentation:**
> [Start only some container](https://blibox.readthedocs.io/en/latest/getting-started/start-the-blibox.html#start-some-container)

![Blibox](docs/img/blibox-dash-selective.png)

#### Run different versions

Every single attachable container comes with many different versions. In order to select the desired version for a container, simply edit the `.env` file and uncomment the version of choice. Any combination is possible.

<table>
  <thead>
    <tr>
      <th>Apache</th>
      <th>Nginx</th>
      <th>PHP</th>
      <th>MySQL</th>
      <th>MariaDB</th>
      <th>Percona</th>
      <th>PgSQL</th>
      <th>Redis</th>
      <th>Memcached</th>
      <th>MongoDB</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><a target="_blank" title="Apache 2.2"       href="https://github.com/blibox/docker-apache-2.2">2.2</a></td>
      <td><a target="_blank" title="Nginx stable"     href="https://github.com/blibox/docker-nginx-stable">stable</a></td>
      <td><a target="_blank" title="PHP 5.2"          href="https://github.com/blibox/docker-php-fpm">5.2</a><sup>[1]</sup></td>
      <td><a target="_blank" title="MySQL 5.5"        href="https://github.com/blibox/docker-mysql">5.5</a></td>
      <td><a target="_blank" title="MariaDB 5.5"      href="https://github.com/blibox/docker-mysql">5.5</a></td>
      <td><a target="_blank" title="PerconaDB 5.5"    href="https://github.com/blibox/docker-mysql">5.5</a></td>
      <td><a target="_blank" title="PgSQL 9.0"        href="https://github.com/docker-library/postgres">9.0</a></td>
      <td><a target="_blank" title="Redis 2.8"        href="https://github.com/docker-library/redis">2.8</a></td>
      <td><a target="_blank" title="Memcached 1.4"    href="https://github.com/docker-library/memcached">1.4</a></td>
      <td><a target="_blank" title="MongoDB 2.8"      href="https://github.com/docker-library/mongo">2.8</a></td>
    </tr>
    <tr>
      <td><a target="_blank" title="Apache 2.4"       href="https://github.com/blibox/docker-apache-2.4">2.4</a></td>
      <td><a target="_blank" title="Nginx mainline"   href="https://github.com/blibox/docker-nginx-mainline">mainline</a></td>
      <td><a target="_blank" title="PHP 5.3"          href="https://github.com/blibox/docker-php-fpm">5.3</a></td>
      <td><a target="_blank" title="MySQL 5.6"        href="https://github.com/blibox/docker-mysql">5.6</a></td>
      <td><a target="_blank" title="MariaDB 10.0"     href="https://github.com/blibox/docker-mysql">10.0</a></td>
      <td><a target="_blank" title="PerconaDB 5.6"    href="https://github.com/blibox/docker-mysql">5.6</a></td>
      <td><a target="_blank" title="PgSQL 9.1"        href="https://github.com/docker-library/postgres">9.1</a></td>
      <td><a target="_blank" title="Redis 3.0"        href="https://github.com/docker-library/redis">3.0</a></td>
      <td><a target="_blank" title="Memcached 1.5"    href="https://github.com/docker-library/memcached">1.5</a></td>
      <td><a target="_blank" title="MongoDB 3.0"      href="https://github.com/docker-library/mongo">3.0</a></td>
    </tr>
    <tr>
      <td></td>
      <td></td>
      <td><a target="_blank" title="PHP 5.4"          href="https://github.com/blibox/docker-php-fpm">5.4</a></td>
      <td><a target="_blank" title="MySQL 5.7"        href="https://github.com/blibox/docker-mysql">5.7</a></td>
      <td><a target="_blank" title="MariaDB 10.1"     href="https://github.com/blibox/docker-mysql">10.1</a></td>
      <td><a target="_blank" title="PerconaDB 5.7"    href="https://github.com/blibox/docker-mysql">5.7</a></td>
      <td><a target="_blank" title="PgSQL 9.2"        href="https://github.com/docker-library/postgres">9.2</a></td>
      <td><a target="_blank" title="Redis 3.2"        href="https://github.com/docker-library/redis">3.2</a></td>
      <td><a target="_blank" title="Memcached latest" href="https://github.com/docker-library/memcached">latest</a></td>
      <td><a target="_blank" title="MongoDB 3.2"      href="https://github.com/docker-library/mongo">3.2</a></td>
    </tr>
    <tr>
      <td></td>
      <td></td>
      <td><a target="_blank" title="PHP 5.5"          href="https://github.com/blibox/docker-php-fpm">5.5</a></td>
      <td><a target="_blank" title="MySQL 8.0"        href="https://github.com/blibox/docker-mysql">8.0</a></td>
      <td><a target="_blank" title="MariaDB 10.2"     href="https://github.com/blibox/docker-mysql">10.2</a></td>
      <td><a target="_blank" title="PerconaDB 8.0"    href="https://github.com/blibox/docker-mysql">8.0</a></td>
      <td><a target="_blank" title="PgSQL 9.3"        href="https://github.com/docker-library/postgres">9.3</a></td>
      <td><a target="_blank" title="Redis 4.0"        href="https://github.com/docker-library/redis">4.0</a></td>
      <td></td>
      <td><a target="_blank" title="MongoDB 3.4"      href="https://github.com/docker-library/mongo">3.4</a></td>
    </tr>
    <tr>
      <td></td>
      <td></td>
      <td><a target="_blank" title="PHP 5.6"          href="https://github.com/blibox/docker-php-fpm">5.6</a></td>
      <td></td>
      <td><a target="_blank" title="MariaDB 10.3"     href="https://github.com/blibox/docker-mysql">10.3</a></td>
      <td></td>
      <td><a target="_blank" title="PgSQL 9.4"        href="https://github.com/docker-library/postgres">9.4</a></td>
      <td><a target="_blank" title="Redis 5.0"        href="https://github.com/docker-library/redis">5.0</a></td>
      <td></td>
      <td><a target="_blank" title="MongoDB 3.6"      href="https://github.com/docker-library/mongo">3.6</a></td>
    </tr>
    <tr>
      <td></td>
      <td></td>
      <td><a target="_blank" title="PHP 7.0"          href="https://github.com/blibox/docker-php-fpm">7.0</a></td>
      <td></td>
      <td><a target="_blank" title="MariaDB 10.4"     href="https://github.com/blibox/docker-mysql">10.4</a></td>
      <td></td>
      <td><a target="_blank" title="PgSQL 9.5"        href="https://github.com/docker-library/postgres">9.5</a></td>
      <td><a target="_blank" title="Redis 6.0"        href="https://github.com/docker-library/redis">6.0</a></td>
      <td></td>
      <td><a target="_blank" title="MongoDB 4.0"      href="https://github.com/docker-library/mongo">4.0</a></td>
    </tr>
    <tr>
      <td></td>
      <td></td>
      <td><a target="_blank" title="PHP 7.1"          href="https://github.com/blibox/docker-php-fpm">7.1</a></td>
      <td></td>
      <td></td>
      <td></td>
      <td><a target="_blank" title="PgSQL 9.6"        href="https://github.com/docker-library/postgres">9.6</a></td>
      <td><a target="_blank" title="Redis latest"     href="https://github.com/docker-library/redis">latest</a></td>
      <td></td>
      <td><a target="_blank" title="MongoDB 4.2"      href="https://github.com/docker-library/mongo">4.2</a></td>
    </tr>
    <tr>
      <td></td>
      <td></td>
      <td><a target="_blank" title="PHP 7.2"          href="https://github.com/blibox/docker-php-fpm">7.2</a></td>
      <td></td>
      <td></td>
      <td></td>
      <td>...</td>
      <td></td>
      <td></td>
      <td><a target="_blank" title="MongoDB latest"   href="https://github.com/docker-library/mongo">latest</a></td>
    </tr>
    <tr>
      <td></td>
      <td></td>
      <td><a target="_blank" title="PHP 7.3"          href="https://github.com/blibox/docker-php-fpm">7.3</a></td>
      <td></td>
      <td></td>
      <td></td>
      <td><a target="_blank" title="PgSQL 12.0"       href="https://github.com/docker-library/postgres">12.0</a></td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <td></td>
      <td></td>
      <td><a target="_blank" title="PHP 7.4"          href="https://github.com/blibox/docker-php-fpm">7.4</a></td>
      <td></td>
      <td></td>
      <td></td>
      <td><a target="_blank" title="PgSQL 12.1"       href="https://github.com/docker-library/postgres">12.1</a></td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <td></td>
      <td></td>
      <td><a target="_blank" title="PHP 8.0"          href="https://github.com/blibox/docker-php-fpm">8.0</a><sup>[2]</sup></td>
      <td></td>
      <td></td>
      <td></td>
      <td><a target="_blank" title="PgSQL latest"     href="https://github.com/docker-library/postgres">latest</a></td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
  </tbody>
</table>

<small><strong><sup>[1]</sup></strong> <strong>PHP 5.2</strong> is available to use, but it is not officially supported. The Blibox intranet does not work with this version as PHP 5.2 does not support namespaces. Furthermore PHP 5.2 does only work with Apache 2.4, Nginx stable and Nginx mainline. It does not work with Apache 2.2. Use at your own risk.</small>

<small><strong><sup>[2]</sup></strong> <strong>PHP 8.0</strong> is an upcoming unreleased versions of PHP, which are directly built out of their [official git branches](https://github.com/php/php-src/) every night to assure you will leverage their latest features.</small>

> **Documentation:**
> [Change container versions](https://blibox.readthedocs.io/en/latest/getting-started/change-container-versions.html)

#### Additional services

Additionally to the default stack, there are a variety of other services that can be easily enabled and started.

<table>
 <thead>
  <tr>
   <th>Python (Flask)</th>
   <th>Blackfire</th>
   <th>ELK</th>
   <th>MailHog</th>
   <th>Ngrok</th>
   <th>RabbitMQ</th>
   <th>Solr</th>
   <th>HAProxy</th>
   <th>Varnish</th>
  </tr>
 </thead>
 <tbody>
  <tr>
   <td><a target="_blank" title="Python 2.7   "    href="https://github.com/blibox/docker-python-flask">2.7</a></td>
   <td><a target="_blank" title="Blackfire 1.8"    href="https://github.com/blackfireio/docker">1.8</a></td>
   <td><a target="_blank" title="ELK stack"        href="https://www.docker.elastic.co">5.x.y</a></td>
   <td><a target="_blank" title="MailHog v1.0.0"   href="https://github.com/mailhog/MailHog">v1.0.0</a></td>
   <td><a target="_blank" title="Ngrok 2.x"        href="https://github.com/blibox/docker-ngrok">2.x</a></td>
   <td><a target="_blank" title="RabbitMQ 3.6"     href="https://github.com/rabbitmq/rabbitmq-server">3.6</a></td>
   <td><a target="_blank" title="Solr 5"           href="https://github.com/apache/lucene-solr">5</a></td>
   <td><a target="_blank" title="HAProxy 1.x"      href="https://github.com/blibox/docker-haproxy">1.x</a></td>
   <td><a target="_blank" title="Varnish 4"        href="https://github.com/blibox/docker-varnish">4</a></td>
  </tr>
  <tr>
   <td>...</td>
   <td>...</td>
   <td><a target="_blank" title="ELK stack"        href="https://www.docker.elastic.co">6.x.y</a></td>
   <td><a target="_blank" title="MailHog latest"   href="https://github.com/mailhog/MailHog">latest</a></td>
   <td></td>
   <td><a target="_blank" title="RabbitMQ 3.7"     href="https://github.com/rabbitmq/rabbitmq-server">3.7</a></td>
   <td><a target="_blank" title="Solr 6"           href="https://github.com/apache/lucene-solr">6</a></td>
   <td></td>
   <td><a target="_blank" title="Varnish 5"        href="https://github.com/blibox/docker-varnish">5</a></td>
  </tr>
  <tr>
   <td><a target="_blank" title="Python 3.7   "    href="https://github.com/blibox/docker-python-flask">3.7</a></td>

CHANGELOG.md
# Changelog

Make sure to have a look at [UPDATING](https://github.com/cytopia/blibox/blob/master/UPDATING.md) to see any required steps for updating
major versions.


## Unreleased


## Release v1.7.0 (2020-03-24)

#### Added
- Python Flask


## Bugfix Release v1.6.3 (2020-03-23)

#### Fixed
- Fixed PHP FPM images
- Fixed cert-gen for HAProxy
- Various spelling errors in documentation
- PostgreSQL startup without a password
#### Added
- [#686](https://github.com/cytopia/blibox/pull/686) Added documentation for ExpressEngine
- New .env var: `PGSQL_HOST_AUTH_METHOD`


## Bugfix Release v1.6.2 (2020-02-06)

#### Added
- [#670](https://github.com/cytopia/blibox/issues/670) Add `phalcon` binary to PHP 7.3 and 7.4
- [#664](https://github.com/cytopia/blibox/issues/664) Add PHP `yaml` extension


## Bugfix Release v1.6.1 (2020-01-05)

#### Fixed
- [#662](https://github.com/cytopia/blibox/issues/662) Update to latest Symfony CLI


## Release v1.6.0 (2020-01-04)

#### Changed
- [#642](https://github.com/cytopia/blibox/issues/642) Make email catch-all configurable
- [#265](https://github.com/cytopia/blibox/issues/265) Make SSL vhost settings configurable

#### Added
- [#615](https://github.com/cytopia/blibox/issues/615) Add phpmd
- [#378](https://github.com/cytopia/blibox/issues/378) Allow to mount local `.ssh/` directory into PHP container (read-only)


## Release v1.5.0 (2020-01-03)

#### Added
- [#654](https://github.com/cytopia/blibox/issues/654) Added Opcache Control Panel
- Integration tests for MySQL Docker image


## Release v1.4.0 (2020-01-02)

#### Fixed
- [#618](https://github.com/cytopia/blibox/issues/618) Update Compose version to 2.3
- [#614](https://github.com/cytopia/blibox/issues/614) Update to latest mhsendmail binary with `-o` flag
- [#265](https://github.com/cytopia/blibox/issues/265) Fix http to https redirect for projects

#### Changed
- [#642](https://github.com/cytopia/blibox/issues/642) Be able to disable email catch-all and still run postfix


## Release v1.3.0 (2019-12-29)

#### Fixed
- [#626](https://github.com/cytopia/blibox/issues/626) Cannot create MongoDB database with Adminer

#### Added
- latest PHP-FPM images
- PHP module: solr
- PHP module: ssh2
- phpMyAdmin to 5.0.0

#### Changed
- Use official PHP 7.4 Docker image as base
- Updated Adminer to 4.7.5
```
sed -i'' 's/^<?php$/<?php if(!function_exists("get_magic_quotes_runtime")){function get_magic_quotes_runtime(){return false;}}if(!function_exists("get_magic_quotes_gpc")){function get_magic_quotes_gpc(){return false;}}/g' adminer-4.7.5-en.php
sed -i'' 's/while(list(\$y,\$X)=each(\$qg))/foreach ($qg as $y => $x)/g' adminer-4.7.5-en.php
sed -i'' 's/error_reporting(6135)/error_reporting(0)/g' adminer-4.7.5-en.php
```
- Updated phpMyAdmin to 4.9.3


## Release v1.2.0 (2019-12-01)

#### Fixed
- [#622](https://github.com/cytopia/blibox/issues/622) Certificate Generation Settings
- [#640](https://github.com/cytopia/blibox/issues/640) Mac OS Catalina invalidates virtual host certificates
- [#592](https://github.com/cytopia/blibox/issues/592) sqlsrv connection problem

#### Added
- Added latest PHP-FPM images
- Added latest MySQL images
- Added latest Apache/Nginx images

#### Changed
- Made PHP 7.3 the default version


## Release v1.1.0 (2019-11-24)

#### Fixed
- [#644](https://github.com/cytopia/blibox/issues/644) Fix GD jpeg support missing in PHP 7.4
- [#619](https://github.com/cytopia/blibox/issues/619) Fix PHP 7.x  WebP support with imagick

#### Added
- Make Ngrok region configurable via `NGROK_REGION` env variable
- [#641](https://github.com/cytopia/blibox/issues/641) Added phpPgAdmin v7 for PHP >=7
- [#594](https://github.com/cytopia/blibox/issues/594) Tool: ghostscript
- Extensive GitHub Action CI checks
- Added new PostgreSQL images
- Added new MongoDB images

#### Changed
- Updated PHP Docker images (PHP version, modules and tools)


## Bugfix Release v1.0.2 (2019-05-21)

#### Fixed
- Fixed various typos in documentation
- Fix CI tests: They still expected a mounted mail directory instead of a Docker volume
- Fix mods for PHP 8.0
- Make npm binaries available in $PATH
- Remove orphaned mentions of HOST_PATH_MYSQL_DATADIR
- Allow symlinks in autostart scripts

#### Added
- Documentation: Setup ProcessWire
- Tool: Angular CLI
- Tool: Laravel Lumen
- Tool: [prestissimo](https://github.com/hirak/prestissimo)
- Tool: [yq](https://github.com/mikefarah/yq)
- Module: OAuth

#### Removed
- Enchant module for PHP 7.4 and PHP 8.0 (build breaks)


## Bugfix Release v1.0.1 (2019-03-24)

This is a bugfix release and everybody is encouraged to upgrade to this tag as soon as possible.
No explicit actions to be taken for updating.

#### Fixed
- [#373](https://github.com/cytopia/blibox/issues/373) Read custom MySQL configuration
- [#540](https://github.com/cytopia/blibox/issues/540) Fix '&' password substitution in mysqldump-secure
- [#209](https://github.com/cytopia/blibox/issues/209) Documentation for Xdebug on Docker for Windows

#### Changed
- Updated MongoDB cli tools in PHP image
- Updated PostgreSQL cli tools in PHP image
- MySQL images are now bound to a specific Docker tag and are built nightly
- [#506](https://github.com/cytopia/blibox/issues/506) Documentation improvements for connecting to databases

#### Added
- [#536](https://github.com/cytopia/blibox/issues/536) Added @vue/cli and @vue/cli-service-global



## Release v1.0.0 (2019-03-19)

This is the first major stable release of the Blibox.

#### Changed
- Everything from v1.0.0-alpha1 has been backported
- Everything from v0.15.0 has been backported



## Pre-Release v1.0.0-alpha1 (2019-03-09)

#### Changed
- Use Docker volumes instead of directory mounts for stateful data (MySQL, PgSQL and MongoDB)
    - This fixes various mount issues on Windows: #175 #382
    - This improves general performance
- Use Official MySQL, MariaDB and Percona Docker container



## Release v0.15.0 (2019-03-09)

This will be the last v0.x release.

#### Fixed

.blibox/www/config.php
<?PHP
// Measure time
$TIME_START = microtime(true);

// Start session
session_start();

// Turn on all PHP errors
error_reporting(-1);


// Shorten DNS timeouts for gethostbyname in case DNS server is down
putenv('RES_OPTIONS=retrans:1 retry:1 timeout:1 attempts:1');


$BLIBOX_VERSION = 'v1.7.0';
$BLIBOX_DATE = '2020-03-24';
$BLIBOX_API_PAGE = 'blibox-api/status.json';

//
// Set Directories
//
$CONF_DIR	= dirname(__FILE__);
$LIB_DIR	= $CONF_DIR . DIRECTORY_SEPARATOR . 'include' . DIRECTORY_SEPARATOR .'lib';
$VEN_DIR	= $CONF_DIR . DIRECTORY_SEPARATOR . 'include' . DIRECTORY_SEPARATOR .'vendor';
$LOG_DIR	= dirname(dirname($CONF_DIR)) . DIRECTORY_SEPARATOR . 'log' . DIRECTORY_SEPARATOR . 'blibox';


//
// Load Base classes
//
require $LIB_DIR . DIRECTORY_SEPARATOR . 'container' . DIRECTORY_SEPARATOR .'BaseClass.php';
require $LIB_DIR . DIRECTORY_SEPARATOR . 'container' . DIRECTORY_SEPARATOR .'BaseInterface.php';



//
// Set Docker addresses
//
$DNS_HOST_NAME		= 'bind';
$PHP_HOST_NAME		= 'php';
$HTTPD_HOST_NAME	= 'httpd';
$MYSQL_HOST_NAME	= 'mysql';
$PGSQL_HOST_NAME	= 'pgsql';
$REDIS_HOST_NAME	= 'redis';
$MEMCD_HOST_NAME	= 'memcd';
$MONGO_HOST_NAME	= 'mongo';


//
// Lazy Container Loader
//
function loadFile($class, $base_path) {
	static $_LOADED_FILE;

	if (isset($_LOADED_FILE[$class])) {
		return;
	}

	require $base_path . DIRECTORY_SEPARATOR . $class . '.php';
	$_LOADED_FILE[$class] = true;
	return;
}
function loadClass($class) {

	static $_LOADED_LIBS;

	if (isset($_LOADED_LIBS[$class])) {
		return $_LOADED_LIBS[$class];
	} else {

		$lib_dir = $GLOBALS['LIB_DIR'];
		$cnt_dir = $GLOBALS['LIB_DIR'] . DIRECTORY_SEPARATOR . 'container';

		switch($class) {
			//
			// Lib Classes
			//
			case 'Logger':
				loadFile($class, $lib_dir);
				$_LOADED_LIBS[$class] = \blibox\Logger::getInstance();
				break;

			case 'Html':
				loadFile($class, $lib_dir);
				$_LOADED_LIBS[$class] = \blibox\Html::getInstance();
				break;

			case 'Helper':
				loadFile($class, $lib_dir);
				$_LOADED_LIBS[$class] = \blibox\Helper::getInstance();
				break;

			//
			// Docker Container Classes
			//
			case 'Php':
				loadFile($class, $cnt_dir);
				$_LOADED_LIBS[$class] = \blibox\Php::getInstance($GLOBALS['PHP_HOST_NAME']);
				break;

			case 'Dns':
				loadFile($class, $cnt_dir);
				$_LOADED_LIBS[$class] = \blibox\Dns::getInstance($GLOBALS['DNS_HOST_NAME']);
				break;

			case 'Httpd':
				loadFile($class, $cnt_dir);
				$_LOADED_LIBS[$class] = \blibox\Httpd::getInstance($GLOBALS['HTTPD_HOST_NAME']);
				break;

			case 'Mysql':
				loadFile($class, $cnt_dir);
				$_LOADED_LIBS[$class] = \blibox\Mysql::getInstance($GLOBALS['MYSQL_HOST_NAME'], array(
					'user' => 'root',
					'pass' => loadClass('Helper')->getEnv('MYSQL_ROOT_PASSWORD')
				));
				break;

			case 'Pgsql':
				loadFile($class, $cnt_dir);
				$_LOADED_LIBS[$class] = \blibox\Pgsql::getInstance($GLOBALS['PGSQL_HOST_NAME'], array(
					'user' => loadClass('Helper')->getEnv('PGSQL_ROOT_USER'),
					'pass' => loadClass('Helper')->getEnv('PGSQL_ROOT_PASSWORD'),
					'db' => 'postgres'
				));
				break;

			case 'Redis':

				// Check if redis is using a password
				$REDIS_ROOT_PASSWORD = '';

				$_REDIS_ARGS = loadClass('Helper')->getEnv('REDIS_ARGS');
				$_REDIS_PASS = preg_split("/--requirepass\s+/",  $_REDIS_ARGS);
				if (is_array($_REDIS_PASS) && count($_REDIS_PASS)) {
					// In case the option is specified multiple times, use the last effective one.
					$_REDIS_PASS = $_REDIS_PASS[count($_REDIS_PASS)-1];
					if (strlen($_REDIS_PASS) > 0) {
						$REDIS_ROOT_PASSWORD = $_REDIS_PASS;
					}
				}

				loadFile($class, $cnt_dir);
				if ($REDIS_ROOT_PASSWORD == '') {
					$_LOADED_LIBS[$class] = \blibox\Redis::getInstance($GLOBALS['REDIS_HOST_NAME']);
				} else {
					$_LOADED_LIBS[$class] = \blibox\Redis::getInstance($GLOBALS['REDIS_HOST_NAME'], array(
						'pass' => $REDIS_ROOT_PASSWORD,
					));
				}
				break;

			case 'Memcd':
				loadFile($class, $cnt_dir);
				$_LOADED_LIBS[$class] = \blibox\Memcd::getInstance($GLOBALS['MEMCD_HOST_NAME']);
				break;

			case 'Mongo':
				loadFile($class, $cnt_dir);
				$_LOADED_LIBS[$class] = \blibox\Mongo::getInstance($GLOBALS['MONGO_HOST_NAME']);
				break;

			// Get optional docker classes
			default:
				// Unknown class
				exit('Class does not exist: '.$class);
		}
		return $_LOADED_LIBS[$class];
	}
}
